services/net/filemonitor/FilemonitorAction.cs

```diff
@@ -0,0 +1,678 @@
+                var matches = Regex.Matches(doc, "<story>(.+?)</story>", RegexOptions.Singleline);
+                var source = "";
+
+                // Iterate over the list of stories and add a new item to the articles list for each story.
+                foreach (Match story in matches)
+                {
+                    var item = new SourceContent();
+                    var preFiltered = story.Groups[1].Value;
+
+                    // Single line mode prevents matching on "\n\n", so replace this with a meaningful field delimiter.
+                    var filtered = preFiltered.Replace("\n\n", Fields.FmsFieldDelim);
+
+                    var papername = GetFmsData(filtered, Fields.Papername, dataSource);
+                    source = string.IsNullOrEmpty(source) ? await GetItemSourceCodeAsync(dataSource, papername) : source;
+                    item.Source = source;
+                    item.Title = GetFmsData(filtered, Fields.Headline, dataSource);
+                    item.Uid = GetFmsData(filtered, Fields.Id, dataSource);
+                    item.Summary = GetFmsData(preFiltered + "<break>", Fields.Story, dataSource);
+                    item.Language = GetFmsData(filtered, Fields.Lang, dataSource);
+                    item.FilePath = document.Key;
+                    item.Abstract = GetFmsData(filtered, Fields.Summary, dataSource);
+                    item.Page = GetFmsData(filtered, Fields.Page, dataSource);
+                    item.Section = GetFmsData(filtered, Fields.Section, dataSource);
+
+                    item.Tags = GetTagList(filtered, dataSource);
+                    item.Authors = GetAuthorList(GetFmsData(filtered, Fields.Author, dataSource));
+                    item.PublishedOn = GetPublishedOn(GetFmsData(filtered, Fields.Date, dataSource), dataSource);
+
+                    articles.Add(item);
+                }
+            }
+            catch (Exception ex)
+            {
+                _logger.LogInformation(ex, $"File contents for data source '{dataSource.Code}' is invalid.");
+                throw;
+            }
+        }
+
+        return articles;
+    }
+
+    /// <summary>
+    /// Get the data source's "code" value. If the data source is selfPublished (relates to only one publication) the code
+    /// of the data source is returned. If the content string attribute "selfPublished" is false, the actual data source for
+    /// the publication is retrieved by the value of the paperName parameter.
+    ///
+    /// The name of each paper is not guaranteed to be unique, so we may have to define multiple data sources for the same
+    /// publication, one for each unique name.
+    /// </summary>
+    /// <param name="dataSource"></param>
+    /// <param name="paperName"></param>
+    /// <returns></returns>
+    /// <exception cref="InvalidOperationException"></exception>
+    private async Task<string> GetItemSourceCodeAsync(DataSourceModel dataSource, string paperName)
```

Comment:
As I said in my notes, I'm open to changing the approach :-)  I'd like to have a chat about how your suggestion should be implemented.

Reply:
schedule a meeting tomorrow.
The basic implementation is add the mentioned key `DataSource.Connection.sources` and extract from it the key=value pairs that provide you a mapping of each paper name to an appropriate code.

Reply:
If a papername doesn't match use a `*=DEFAULT SOURCE CODE` so that it does get imported.  Probably should log as a `LogWarning` to track how often this is occurring.